libs/ui/src/hooks/smartcard_auth/use_dipped_smartcard_auth.ts

```diff
@@ -0,0 +1,250 @@
+  | { type: 'log_out' }
+  | { type: 'bootstrap_admin_session'; electionHash: string };
+
+function validateCardUser(
+  card: CardApiReady
+): Result<AdminUser | SuperadminUser, LoggedOut['reason']> {
+  const user = parseUserFromCard(card);
+  if (!user) return err('invalid_user_on_card');
+  if (!(user.role === 'admin' || user.role === 'superadmin')) {
+    return err('user_role_not_allowed');
+  }
+  return ok(user);
+}
+
+function smartcardAuthReducer(
+  previousState: SmartcardAuthState,
+  action: SmartcardAuthAction
+): SmartcardAuthState {
+  switch (action.type) {
+    case 'card_read': {
+      const newAuth = ((): AuthState => {
+        switch (previousState.auth.status) {
+          case 'logged_out': {
+            switch (action.card.status) {
+              case 'no_card':
+                return { status: 'logged_out', reason: 'machine_locked' };
+
+              case 'error':
+                return { status: 'logged_out', reason: 'card_error' };
+
+              case 'ready': {
+                const userResult = validateCardUser(action.card);
+                if (userResult.isOk()) {
+                  const user = userResult.ok();
+                  // TODO: This case can be removed once superadmin cards have passcodes
+                  if (user.role === 'superadmin') {
+                    return { status: 'remove_card', user };
+                  }
+                  return { status: 'checking_passcode', user };
+                }
+                return { status: 'logged_out', reason: userResult.err() };
+              }
+
+              /* istanbul ignore next - compile time check for completeness */
+              default:
+                return throwIllegalValue(action.card, 'status');
+            }
+          }
+
+          case 'checking_passcode': {
+            if (action.card.status === 'no_card') {
+              return { status: 'logged_out', reason: 'machine_locked' };
+            }
+            return previousState.auth;
+          }
+
+          case 'remove_card': {
+            if (action.card.status === 'no_card') {
+              return { status: 'logged_in', user: previousState.auth.user };
+            }
+            return previousState.auth;
+          }
+
+          case 'logged_in':
+            return previousState.auth;
+
+          /* istanbul ignore next - compile time check for completeness */
+          default:
+            throwIllegalValue(previousState.auth, 'status');
+        }
+      })();
+
+      // Optimization: if the card and auth state didn't change, then we can
+      // return the previous state, which will cause React to not rerender.
+      // https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch
+      const newState: SmartcardAuthState = {
+        auth: newAuth,
+        card: action.card,
+      };
+      return deepEqual(newState, previousState) ? previousState : newState;
+    }
+
+    case 'check_passcode': {
```

Comment:
One potential gap I'm seeing is that we aren't checking PINs for admins in the hook for inserted card auth ([something that we do want](https://docs.google.com/document/d/1jBoQqIE7IEr3G_zGfAkl0gDD2kammjIDzpOf7K7StsI/edit#heading=h.efhr9z9u98rx)) - are you planning to add that in later?

Reply:
I was just replicating existing behavior when I implemented inserted card auth for the BMD, which didn't check admin card passcodes. Since this PR doesn't modify inserted card auth at all, I think it should be handled in a follow up. It wasn't on my personal roadmap for these refactors.

Reply:
Got it - I agree that matching existing behavior exactly makes the most sense for this PR!